File: matlab/nii2png.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%       nii2png for Matlab R2017b       %
%         NIfTI Image Converter         %
%                v0.2.9                 %
%                                       %
%     Written by Alexander Laurence     %
% http://Celestial.Tokyo/~AlexLaurence/ %
%    alexander.adamlaurence@gmail.com   %
%              09 May 2019              %
%              MIT License              %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Set the Working Directory
clear; warning off; current = pwd;
path = uigetdir(pwd,'select your working directory');
cd(path);

% Select NIfTI
[file,path] = uigetfile('*.nii');
if isequal(file,0)
   disp('User selected Cancel');
else
   disp(['User selected ', fullfile(path,file)]);
   
   % Read NIfTI Data and Header Info
   image = niftiread(fullfile(path,file));
   image_info = niftiinfo(fullfile(path,file));
   nifti_array = size(image);
   double = im2double(image);
   
   % ask user to rotate and by how much
   ask_rotate = input(' Would you like to rotate the orientation? (y/n) ', 's');
   if lower(ask_rotate) == 'y'
       ask_rotate_num = str2double(input('OK. By 90° 180° or 270°? ', 's'));
       if ask_rotate_num == 90 || ask_rotate_num == 180 || ask_rotate_num == 270
           disp('Got it. Your images will be rotated.');
       else
           disp('Sorry, I did not understand that. Quitting...');
           exit;
       end
   elseif lower(ask_rotate) ~= 'n' && lower(ask_rotate) ~= 'y'
       disp('Sorry, I did not understand that. Quitting...');
       exit;
   end

   % If this is a 4D NIfTI
   if length(nifti_array) == 4
       
       % Create output folder
       mkdir png
       
       % Get Vols and Slice
       total_volumes = nifti_array(4);
       total_slices = nifti_array(3);
       
       current_volume = 1;
       disp('Converting NIfTI to png, please wait...')
       % Iterate Through Vol
       while current_volume <= total_volumes
           slice_counter = 0;
            % Iterate Through Slices
            current_slice = 1;
            while current_slice <= total_slices
                % Alternate Slices
                if mod(slice_counter, 1) == 0
                    
                    % Rotate images if selected
                    if lower(ask_rotate) == 'y'
                        if ask_rotate_num == 90
                            data = rot90(mat2gray(double(:,:,current_slice,current_volume)));
                        elseif ask_rotate_num == 180
                            data = rot90(rot90(mat2gray(double(:,:,current_slice,current_volume))));
                        elseif ask_rotate_num == 270
                            data = rot90(rot90(rot90(mat2gray(double(:,:,current_slice,current_volume)))));
                        end
                    elseif lower(ask_rotate) == 'n'
                    disp('OK, I will convert it as it is.');
                    data = mat2gray(double(:,:,current_slice,current_volume));
                    end
                    
                    % Set Filename as per slice and vol info
                    filename = file(1:end-4) + "_t" + sprintf('%03d', current_volume) + "_z" + sprintf('%03d', current_slice) + ".png";
                    
                    % Write Image
                    imwrite(data, char(filename));
                                        
                    % If we reached the end of the slices
                    if current_slice == total_slices
                        % But not the end of the volumes
                        if current_volume < total_volumes
                            % Move to the next volume                            
                            current_volume = current_volume + 1;
                            % Write the image
                            imwrite(data, char(filename));
                        % Else if we reached the end of slice and volume
                        else         
                            % Write Image
                            imwrite(data, char(filename));
                            disp('Finished!')
                            return
                        end
                    end
                 
                    % Move Images To Folder
                    movefile(char(filename),'png');
                    
                    % Increment Counters
                    slice_counter = slice_counter + 1;
                    
                    percentage = strcat('Please wait. Converting...', ' ', num2str((current_volume/total_volumes)*100), '% Complete');
                    
                    if ((current_volume/total_volumes)*100) == 100
                        disp('100% Complete! Images successfully converted.');
                    else
                        disp(percentage);
                    end                    
                end
                current_slice = current_slice + 1;
            end
       current_volume = current_volume + 1;
       end
   % Else if this is a 3D NIfTI
   elseif length(nifti_array) == 3
       % Create output folder
       mkdir png
       
       % Get Vols and Slice
       total_slices = nifti_array(3);
       
       disp('Converting NIfTI to png, please wait...')

       slice_counter = 0;
        % Iterate Through Slices
        current_slice = 1;
        while current_slice <= total_slices
            % Alternate Slices
            if mod(slice_counter, 1) == 0     
                
                % Rotate images if selected
                if lower(ask_rotate) == 'y'
                    if ask_rotate_num == 90
                        data = rot90(mat2gray(double(:,:,current_slice)));
                    elseif ask_rotate_num == 180
                        data = rot90(rot90(mat2gray(double(:,:,current_slice))));
                    elseif ask_rotate_num == 270
                        data = rot90(rot90(rot90(mat2gray(double(:,:,current_slice)))));
                    end
                elseif lower(ask_rotate) == 'n'
                disp('OK, I will convert it as it is.');
                data = mat2gray(double(:,:,current_slice));
                end
                
                % Set Filename as per slice and vol info
                filename = file(1:end-4) + "_z" + sprintf('%03d', current_slice) + ".png";

                % Write Image
                imwrite(data, char(filename));

                % Move Images To Folder
                movefile(char(filename),'png');

                % Increment Counters
                slice_counter = slice_counter + 1;

                percentage = strcat('Please wait. Converting...', ' ', num2str((current_slice/total_slices)*100), '% Complete');

                if ((current_slice/total_slices)*100) == 100
                    disp('100% Complete! Images successfully converted');
                else
                    disp(percentage);
                end                    
            end
            current_slice  = current_slice  + 1;
        end
   elseif length(nifti_array) ~= 3 || 4
       disp('NIfTI must be 3D or 4D. Please try again.');
   end
end

File: python/nii2png.py
#!/usr/bin/env python
#########################################
#       nii2png for Python 3.7          #
#         NIfTI Image Converter         #
#                v0.2.9                 #
#                                       #
#     Written by Alexander Laurence     #
# http://Celestial.Tokyo/~AlexLaurence/ #
#    alexander.adamlaurence@gmail.com   #
#              09 May 2019              #
#              MIT License              #
#########################################

import numpy, os, nibabel
import sys, getopt
from matplotlib import image

def main(argv):
    inputfile = ''
    outputdir = ''
    try:
        opts, args = getopt.getopt(argv,"hi:o:",["ifile=","ofile="])
    except getopt.GetoptError:
        print('nii2png.py -i <inputfile> -o <outputfile>')
        sys.exit(2)
    for opt, arg in opts:
        if opt == '-h':
            print('nii2png.py -i <inputfile> -o <outputfile>')
            sys.exit()
        elif opt in ("-i", "--input"):
            inputfile = arg
        elif opt in ("-o", "--output"):
            outputdir = arg

    print('Input file is ', inputfile)
    print('Output folder is ', outputdir)

    # set fn as your 4d nifti file
    image_nib: nibabel.Nifti1Image = nibabel.load(inputfile)
    image_array = image_nib.get_fdata()
    print(len(image_array.shape))

    # ask if rotate
    ask_rotate = input('Would you like to rotate the orientation? (y/n) ')

    if ask_rotate.lower() == 'y':
        ask_rotate_num = int(input('OK. By 90° 180° or 270°? '))
        if ask_rotate_num == 90 or ask_rotate_num == 180 or ask_rotate_num == 270:
            print('Got it. Your images will be rotated by {} degrees.'.format(ask_rotate_num))
        else:
            print('You must enter a value that is either 90, 180, or 270. Quitting...')
            sys.exit()
    elif ask_rotate.lower() == 'n':
        print('OK, Your images will be converted it as it is.')
    else:
        print('You must choose either y or n. Quitting...')
        sys.exit()

    # if 4D image inputted
    if len(image_array.shape) == 4:
        # set 4d array dimension values
        nx, ny, nz, nw = image_array.shape

        # set destination folder
        if not os.path.exists(outputdir):
            os.makedirs(outputdir)
            print("Created ouput directory: " + outputdir)

        print('Reading NIfTI file...')

        total_volumes = image_array.shape[3]
        total_slices = image_array.shape[2]

        # iterate through volumes
        for current_volume in range(0, total_volumes):
            slice_counter = 0
            # iterate through slices
            for current_slice in range(0, total_slices):
                if (slice_counter % 1) == 0:
                    # rotate or no rotate
                    if ask_rotate.lower() == 'y':
                        if ask_rotate_num == 90 or ask_rotate_num == 180 or ask_rotate_num == 270:
                            print('Rotating image...')
                            if ask_rotate_num == 90:
                                data = numpy.rot90(image_array[:, :, current_slice, current_volume])
                            elif ask_rotate_num == 180:
                                data = numpy.rot90(numpy.rot90(image_array[:, :, current_slice, current_volume]))
                            elif ask_rotate_num == 270:
                                data = numpy.rot90(numpy.rot90(numpy.rot90(image_array[:, :, current_slice, current_volume])))
                    elif ask_rotate.lower() == 'n':
                        data = image_array[:, :, current_slice, current_volume]
                            
                    #alternate slices and save as png
                    print('Saving image...')
                    image_name = os.path.basename(inputfile).split('.', 1)[0] + "_z" + "{:0>3}".format(str(current_slice+1))+ ".png"
                    image.imsave(os.path.join(outputdir, image_name), data, cmap='gray')
                    slice_counter += 1
                    print('Saved.')

        print('Finished converting images')

    # else if 3D image inputted
    elif len(image_array.shape) == 3:
        # set 4d array dimension values
        nx, ny, nz = image_array.shape

        # set destination folder
        if not os.path.exists(outputdir):
            os.makedirs(outputdir)
            print("Created ouput directory: " + outputdir)

        print('Reading NIfTI file...')

        total_slices = image_array.shape[2]

        slice_counter = 0
        # iterate through slices
        for current_slice in range(0, total_slices):
            # alternate slices
            if (slice_counter % 1) == 0:
                # rotate or no rotate
                if ask_rotate.lower() == 'y':
                    if ask_rotate_num == 90 or ask_rotate_num == 180 or ask_rotate_num == 270:
                        if ask_rotate_num == 90:
                            data = numpy.rot90(image_array[:, :, current_slice])
                        elif ask_rotate_num == 180:
                            data = numpy.rot90(numpy.rot90(image_array[:, :, current_slice]))
                        elif ask_rotate_num == 270:
                            data = numpy.rot90(numpy.rot90(numpy.rot90(image_array[:, :, current_slice])))
                elif ask_rotate.lower() == 'n':
                    data = image_array[:, :, current_slice]

                #alternate slices and save as png
                if (slice_counter % 1) == 0:
                    print('Saving image...')
                    image_name = os.path.basename(inputfile).split('.', 1)[0] + "_z" + "{:0>3}".format(str(current_slice+1))+ ".png"
                    image.imsave(os.path.join(outputdir, image_name), data, cmap='gray')
                    slice_counter += 1
                    print('Saved.')

        print('Finished converting images')
    else:
        print('Not a 3D or 4D Image. Please try again.')

# call the function to start the program
if __name__ == "__main__":
   main(sys.argv[1:])
